--- a/ref/matrixmod.c
+++ b/ref/matrixmod.c
@@ -129,19 +129,19 @@
     // swap
     for (int r2 = r+1; r2 < M_r; r2++)
     {
-      uint64_t Mrr = pmod_mat_entry(M, M_r, M_c, r, r);
+      uint32_t Mrr = pmod_mat_entry(M, M_r, M_c, r, r);
 
       for (int c = r; c < M_c; c++)
       {
-        uint64_t val = pmod_mat_entry(M, M_r, M_c, r2, c);
-
-        uint64_t Mrc = pmod_mat_entry(M, M_r, M_c, r, c);
+        uint32_t val = pmod_mat_entry(M, M_r, M_c, r2, c);
+
+        uint32_t Mrc = pmod_mat_entry(M, M_r, M_c, r, c);
 
         pmod_mat_set_entry(M, M_r, M_c, r, c, (Mrc + val * (Mrr == 0)) % MEDS_p);
       }
     }
 
-    uint64_t val = pmod_mat_entry(M, M_r, M_c, r, r);
+    uint32_t val = pmod_mat_entry(M, M_r, M_c, r, r);
 
     if (val == 0)
       return -1;
@@ -151,21 +151,21 @@
     // normalize
     for (int c = r; c < M_c; c++)
     {
-      uint64_t tmp = ((uint64_t)pmod_mat_entry(M, M_r, M_c, r, c) * val) % MEDS_p;
+      uint32_t tmp = ((uint32_t)pmod_mat_entry(M, M_r, M_c, r, c) * val) % MEDS_p;
       pmod_mat_set_entry(M, M_r, M_c, r, c, tmp);
     }
 
     // eliminate
     for (int r2 = r+1; r2 < M_r; r2++)
     {
-      uint64_t factor = pmod_mat_entry(M, M_r, M_c, r2, r);
+      uint32_t factor = pmod_mat_entry(M, M_r, M_c, r2, r);
 
       for (int c = r; c < M_c; c++)
       {
-        uint64_t tmp0 = pmod_mat_entry(M, M_r, M_c, r, c);
-        uint64_t tmp1 = pmod_mat_entry(M, M_r, M_c, r2, c);
-
-        int64_t val = (tmp0 * factor) % MEDS_p;
+        uint32_t tmp0 = pmod_mat_entry(M, M_r, M_c, r, c);
+        uint32_t tmp1 = pmod_mat_entry(M, M_r, M_c, r2, c);
+
+        int32_t val = (tmp0 * factor) % MEDS_p;
 
         val = tmp1 - val;
 
@@ -185,12 +185,12 @@
   for (int r = M_r - 1; r >= 0; r--)
     for (int r2 = 0; r2 < r; r2++)
     {
-      uint64_t factor = pmod_mat_entry(M, M_r, M_c, r2, r);
-
-      uint64_t tmp0 = pmod_mat_entry(M, M_r, M_c, r, r);
-      uint64_t tmp1 = pmod_mat_entry(M, M_r, M_c, r2, r);
-
-      int64_t val = (tmp0 * factor) % MEDS_p;
+      uint32_t factor = pmod_mat_entry(M, M_r, M_c, r2, r);
+
+      uint32_t tmp0 = pmod_mat_entry(M, M_r, M_c, r, r);
+      uint32_t tmp1 = pmod_mat_entry(M, M_r, M_c, r2, r);
+
+      int32_t val = (tmp0 * factor) % MEDS_p;
 
       val = tmp1 - val;
 
@@ -200,8 +200,8 @@
 
       for (int c = M_r; c < M_c; c++)
       {
-        uint64_t tmp0 = pmod_mat_entry(M, M_r, M_c, r, c);
-        uint64_t tmp1 = pmod_mat_entry(M, M_r, M_c, r2, c);
+        uint32_t tmp0 = pmod_mat_entry(M, M_r, M_c, r, c);
+        uint32_t tmp1 = pmod_mat_entry(M, M_r, M_c, r2, c);
 
         int val = (tmp0 * factor) % MEDS_p;
 
